### PR TITLE
Block data implementation

### PR DESCRIPTION
**Pull request by lucdon**
# Block Data Manipulator Implementation: 
this pr is for implementing all block data catalog type related manipulators 

## manipulator checklist:
- [x] `DirtData`
- [x] `StoneData`
- [x] `BrickData`
- [x] `SandData`
- [x] `SandstoneData`
- [x] `QuartzData`
- [x] `PrismarineData`
- [x] `slabData`
- [x] `ComparatorData`
- [x] `DisguisedBlockData`
- [x] `HingeData`
- [x] `PistonData`
- [x] `PortionData`
- [x] `RailDirectionData`
- [x] `StairShapeData`
- [x] `WallTypeData`
- [x] `ShrubData`
- [x] `PlantData`
- [x] `DoublePlantData`
- [x] `TreeData`
- [x] `BigMushroomData`

## Important: 
* making sure all mixins are registered correctly 
* making sure all types are registered correctly

## implementation checkList:
- [x] implemented all BlockMixins
- [x] implemented all EnumTypeMixins
- [x] implemented all DataProcessors
- [x] implemented all ValueProcessors
- [x] registered all CatalogTypes
- [x] registered all BlockMixins
- [x] registered all EnumTypeMixins
- [x] registered all keys
- [x] registered all implementations

## Test:
https://gist.github.com/lucdon/519108823265505ad4d7
### Results:
https://gist.github.com/lucdon/0bb477751a7fc1efbe59

### CODE REVIEW

File: src/main/java/org/spongepowered/common/data/processor/common/AbstractCatalogDataProcessor.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.data.value.mutable.Value;
+import org.spongepowered.common.Sponge;
+import org.spongepowered.common.data.util.DataUtil;
+
+import java.util.Optional;
+import java.util.function.Predicate;
+
+public abstract class AbstractCatalogDataProcessor<T, V extends BaseValue<T>, M extends DataManipulator<M, I>, I extends ImmutableDataManipulator<I, M>>
+        extends AbstractSingleDataProcessor<T, V, M, I> {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/common/AbstractCatalogDataProcessor.java]: If you're basing this off of an `ItemStack`, why not extend `AbstractItemSingleDataProcessor`?

**Comment by gabizou** [comment]
@lucdon would you be able to hop on IRC in #spongedev ?

File: src/main/java/org/spongepowered/common/data/processor/common/AbstractCatalogDataProcessor.java (hunk 1 above)
Comment (lucdon) [inline src/main/java/org/spongepowered/common/data/processor/common/AbstractCatalogDataProcessor.java]: i am not sure if all the blockcatalogtypes can be implemented by using `AbstractItemSingleDataProcessor`. now i can easily change things later on

Comment (Zidane) [inline src/main/java/org/spongepowered/common/data/processor/common/AbstractCatalogDataProcessor.java]: @lucdon 

Be sure to check your email that you used to sign up for your github account with.

File: src/main/java/org/spongepowered/common/data/processor/common/AbstractBlockOnlyCatalogDataProcessor.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.spongepowered.api.data.key.Key;
+import org.spongepowered.api.data.manipulator.DataManipulator;
+import org.spongepowered.api.data.manipulator.ImmutableDataManipulator;
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+
+import java.util.Optional;
+
+public abstract class AbstractBlockOnlyCatalogDataProcessor<T, V extends BaseValue<T>, M extends DataManipulator<M, I>, I extends ImmutableDataManipulator<I, M>>
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/processor/common/AbstractBlockOnlyCatalogDataProcessor.java]: Wouldn't you want to set `T extends CatalogType`?

**Comment by lucdon** [comment]
@gabizou  good luck reviewing :)

**Comment by gabizou** [comment]
@lucdon, if you can revise the commits into one commit?

**Comment by gabizou** [comment]
Please rebase/squash the PR into fewer commits. As is it is a brain ache to get the rebase right.

**Comment by lucdon** [comment]
@gabizou wow it exactly worked :)

**Pull request closed without merging**